### PR TITLE
Implement https://fixer.io as rate provider

### PR DESCRIPTION
**Pull request by jbrinksmeier**
Due to yahoo's outage since yesterday, I implemented https://fixer.io as rate provider

### CODE REVIEW

**Comment by jbrinksmeier** [comment]
looks like the mocking library (https://github.com/mockery/mockery) I introduced is not compatible with php < 5.6 
Is this relevant? shall I rollback to phpunit's mocking capability?

File: src/Provider/Adapter/AdapterInterface.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+interface AdapterInterface
+{
+    /**
+     * Fetches file from path or url, returns raw string contents
+     *
+     * @param string $pathOrUrl
+     * @return string
+     */
+    public function getFile($pathOrUrl);
+}
```
Comment (surajprogrammez) [inline src/Provider/Adapter/AdapterInterface.php]: New line please :)

File: src/Provider/Adapter/CurlAdapter.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
+<?php
+namespace CurrencyConverter\Provider\Adapter;
+
+/**
+ * Fetch files with cUrl
+ *
+ * @package CurrencyConverter\Provider\Adapter
+ */
+class CurlAdapter implements AdapterInterface
```
Comment (surajprogrammez) [inline src/Provider/Adapter/CurlAdapter.php]: You can also use guzzle and create an API adapter

File: src/Provider/Adapter/CurlAdapter.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        curl_setopt($ch, CURLOPT_URL, $pathOrUrl);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1)');
+        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, $timeout);
+        $rawdata = curl_exec($ch);
+        curl_close($ch);
+
+        return $rawdata;
+    }
+}
```
Comment (surajprogrammez) [inline src/Provider/Adapter/CurlAdapter.php]: New line please :)

File: src/Provider/FixerApi.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
+<?php
+namespace CurrencyConverter\Provider;
+
+use CurrencyConverter\Provider\Adapter\AdapterInterface;
+use CurrencyConverter\Provider\Adapter\CurlAdapter;
+
+/**
+ * Fetch rates from https://fixer.io
+ *
+ * @package CurrencyConverter\Provider
```
Comment (surajprogrammez) [inline src/Provider/FixerApi.php]: Not needed as it doesn't add any semantic meaning over the code

File: src/Provider/FixerApi.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        if (null === $this->adapter) {
+            $this->setAdapter(new CurlAdapter());
+        }
+        return $this->adapter;
+    }
+
+    /**
+     * @param AdapterInterface $adapter
+     */
+    public function setAdapter(AdapterInterface $adapter)
```
Comment (surajprogrammez) [inline src/Provider/FixerApi.php]: Use constructor

File: src/Provider/FixerApi.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        $this->adapter = $adapter;
+    }
+}
```
Comment (surajprogrammez) [inline src/Provider/FixerApi.php]: New line please :)

File: src/Provider/FixerApi.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+     * @inheritdoc
+     */
+    public function getRate($fromCurrency, $toCurrency)
+    {
+        $path = sprintf(
+            self::FIXER_API_BASEPATH . '?symbols=%s&base=%s',
+            $toCurrency,
+            $fromCurrency
+        );
+        $result = json_decode($this->getAdapter()->getFile($path));
```
Comment (surajprogrammez) [inline src/Provider/FixerApi.php]: Associative array may be? object isn't suitable here

File: tests/CurrencyConverterTest/Provider/FixerApiTest.php [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+class FixerApiTest extends \PHPUnit_Framework_TestCase
+{
+    protected function tearDown()
+    {
+        m::close();
+    }
+
+    /**
+     * test the default adapter is CurlAdapter
```
Comment (surajprogrammez) [inline tests/CurrencyConverterTest/Provider/FixerApiTest.php]: No need to put doc block here. Your test should describe the purpose.

File: tests/CurrencyConverterTest/Provider/FixerApiTest.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
+<?php
+namespace CurrencyConverter\Provider;
+
+use CurrencyConverter\Provider\Adapter\AdapterInterface;
+use Mockery as m;
```
Comment (surajprogrammez) [inline tests/CurrencyConverterTest/Provider/FixerApiTest.php]: Put it up

File: tests/CurrencyConverterTest/Provider/FixerApiTest.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    public function testDefaultAdapter()
+    {
+        $fixer = new FixerApi();
+        $this->assertInstanceOf(CurlAdapter::class, $fixer->getAdapter());
+    }
+
+    public function testGetRate()
+    {
+        $from = 'EUR';
```
Comment (surajprogrammez) [inline tests/CurrencyConverterTest/Provider/FixerApiTest.php]: If you aren't using it more than once there's no point in creating variables here.

File: tests/CurrencyConverterTest/Provider/FixerApiTest.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        m::close();
+    }
+
+    /**
+     * test the default adapter is CurlAdapter
+     */
+    public function testDefaultAdapter()
+    {
+        $fixer = new FixerApi();
```
Comment (surajprogrammez) [inline tests/CurrencyConverterTest/Provider/FixerApiTest.php]: Not needed you can instantiate it directly in assertion.

File: tests/CurrencyConverterTest/Provider/YahooApiTest.php [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+namespace CurrencyConverter\Provider;
+
+
+use CurrencyConverter\Provider\Adapter\CurlAdapter;
+
+class YahooApiTest extends \PHPUnit_Framework_TestCase
+{
+    public function testDefaultAdapter()
+    {
+        $yahoo = new YahooApi();
```
Comment (surajprogrammez) [inline tests/CurrencyConverterTest/Provider/YahooApiTest.php]: Same as above no need to declare as a variable, you can use it in assertion.

**Comment by surajprogrammez** [comment]
@jbrinksmeier please use PHPUnit mock.

File: src/Provider/FixerApi.php (hunk 6 above)
Comment (surajprogrammez) [inline src/Provider/FixerApi.php]: New line please :)

File: src/Provider/YahooApi.php [hunk 13]
```diff
@@ -23,16 +30,29 @@ public function getRate($fromCurrency, $toCurrency)
...
+    }
+
+    /**
+     * @param CurlAdapter $adapter
+     */
+    public function setAdapter(CurlAdapter $adapter)
+    {
+        $this->adapter = $adapter;
+    }
+
```
Comment (surajprogrammez) [inline src/Provider/YahooApi.php]: Extra space ;) check down as well

**Comment by surajprogrammez** [comment]
@jbrinksmeier when you're done please rebase

**Comment by jbrinksmeier** [comment]
I applied to your semantics, but there are 2 things I would like to discuss:
 - please confirm that you're ok with introducing guzzle as a dependency (wanted to avoid that, hence the "Adapter")
- regarding your comment on using constructor injection: very well, but the readme states a zero-config instantiation. shall I keep that by using null as a default?

Something general: following this [thread](https://yahoo.uservoice.com/forums/382977-finance/suggestions/32103877-http-download-finance-yahoo-com-d-quotes-csv-s-a), yahoo discontinued their API entirely, so maybe we should remove it from this library? 

And, last but not least, I'm afraid I need some guidance on "rebase". I'm familiar with the term, but miss context here as I don't do PRs regularly :)

File: src/Provider/FixerApi.php (hunk 5 above)
Comment (ojhaujjwal) [reply src/Provider/FixerApi.php]: constructor whenever possible

File: src/Provider/FixerApi.php [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+        return $result->rates->$toCurrency;
+    }
+
+    /**
+     * @return AdapterInterface
+     */
+    public function getAdapter()
```
Comment (ojhaujjwal) [inline src/Provider/FixerApi.php]: a public method is like a child. once, you have created it, you have to maintain it for the rest of the lifetime ;) 
you can mark it as private.

**Comment by ojhaujjwal** [comment]
thank you very much for the PR.

**Comment by ojhaujjwal** [comment]
> please confirm that you're ok with introducing guzzle as a dependency

I think guzzle as a dependency shouldn't be a problem.

> regarding your comment on using constructor injection: very well, but the readme states a zero-config instantiation. shall I keep that by using null as a default?
no point breaking BC for this fairly simple library. so you can expect null in the constructor and then set a default value in the constructor.

**Comment by ojhaujjwal** [comment]
Can you please speed it up a little bit?
If you are busy, is it okay if I can take over from you?

**Comment by jbrinksmeier** [comment]
hey there. will finish this by end of today. sorry for the delay, have been busy indeed. fyi: guzzle is implemented now, and i will remove the yahoo code entirely as the service is discontinued.

**Comment by ojhaujjwal** [comment]
Awesome man. Thanks for the update

**Comment by jbrinksmeier** [comment]
hi there, requested changes are ready in this branch: https://github.com/jbrinksmeier/currency-converter-php/tree/PR-11

Please excuse, again, the delay. These old-style phpunit4 mocks gave me a bit of a hard time^^

Shall I open a new PR with that branch? Or how should I continue here?

**Comment by ojhaujjwal** [comment]
you can create a new PR man.

**Pull request merged**